tests/lib/rules/v-on-handler-style.js

```diff
@@ -0,0 +1,913 @@
+/**
+ * @author Yosuke Ota <https://github.com/ota-meshi>
+ * See LICENSE file in root directory for full license.
+ */
+'use strict'
+
+const RuleTester = require('eslint').RuleTester
+const rule = require('../../../lib/rules/v-on-handler-style')
+
+const tester = new RuleTester({
+  parser: require.resolve('vue-eslint-parser'),
+  parserOptions: { ecmaVersion: 2020, sourceType: 'module' }
+})
+
+tester.run('v-on-handler-style', rule, {
+  valid: [
+    {
+      filename: 'test.vue',
+      code: ''
+    },
+    {
+      filename: 'test.vue',
+      code: `<template>
+        <button @click="foo" />
+        <button @click="() => foo()" />
+      </template>`,
+      options: [['method', 'inline-function']]
+    },
+    {
+      filename: 'test.vue',
+      code: `<template>
+        <button @click="foo" />
+        <button @click="() => foo()" />
+      </template>`
+    },
+    {
+      filename: 'test.vue',
+      code: `<template>
+        <button @click="foo()" />
+      </template>`,
+      options: [['inline']]
+    },
+    {
+      filename: 'test.vue',
+      code: '<template><button @[foo()]="bar" /></template>'
+    },
+    {
+      filename: 'test.vue',
+      code: '<template><button :click="foo()" /></template>'
+    },
+    // inline -> method
+    {
+      filename: 'test.vue',
+      code: `<template>
+        <button @click="foo(123)" />
+        <button @click="foo.bar()" />
+        <button @click="foo?.()" />
+        <button @click="foo();foo();" />
+        <button @click="{}" />
```

Comment:
Shouldn't these cases be reported (without fix) to suggest moving the code into methods?

Reply:
We need to check if the expression other than the identifier call expression can be replaced with method + method handler.
For example, if you're using a variable defined with v-for, we can't replace it with a method handler.

```vue
<template v-for="e in list">
  <button @click="handler(e)" />
  <button @click="handlers[e]()" />
  <button @click="handler(a(b), c(d), e + f)" />
</template>
```

I thought it would be difficult to check it, but now that I've looked into it, it seems I can check it by checking the variable associated with the VElement.

I would change this PR to report inline handlers that can be replaced with method handlers.

Reply:
Hmm. Method style is insufficient as you say.
Do you think the user should be able to specify an option which styles to allow next in case the conversion is not possible for some reason if only method styles are allowed?

Reply:
It's still under discussion, but I've made the changes mentioned in https://github.com/vuejs/eslint-plugin-vue/pull/2009#discussion_r993185072.

Reply:
What do you think about adding options like:

```json
{
  "vue/v-on-handler-style": ["error",
    [
      "method",
      "inline-function"
    ],
    {
      "priority": false
    }
  ]
}
```

- priority: true

```vue
<template>
  <!-- ✓ GOOD -->
  <button v-on:click="handler" /> <!-- method handler -->
  <template v-for="e in list">
    <button @click="e" /> <!-- method handler -->
    <button @click="()=>foo(e)" /> <!-- inline function: cannot use method handler -->
  </template>

  <!-- ✗ BAD -->
  <button v-on:click="() => handler()" /> <!-- inline function: can use high priority method handler instead -->
  <template v-for="e in list">
    <button @click="()=>e()" /> <!-- inline function: can use high priority method handler instead -->
  </template>

  <button v-on:click="handler()" /> <!-- inline handler: can use method handler, or inline function instead -->
  <template v-for="e in list">
    <button @click="foo(e)" /> <!-- inline handler: cannot use method handler, but can use inline function instead -->
  </template>
</template>
```

- priority: false (default)

```vue
<template>
  <!-- ✓ GOOD -->
  <button v-on:click="handler" /> <!-- method handler -->
  <button v-on:click="() => handler()" /> <!-- inline function -->
  <template v-for="e in list">
    <button @click="e" /> <!-- method handler -->
    <button @click="()=>foo(e)" /> <!-- inline function -->
    <button @click="()=>e()" /> <!-- inline function -->
  </template>

  <!-- ✗ BAD -->
  <button v-on:click="handler()" /> <!-- inline handler: can use method handler, or inline function instead -->
  <template v-for="e in list">
    <button @click="foo(e)" /> <!-- inline handler: cannot use method handler, but can use inline function -->
  </template>
</template>
```

Reply:
I like the idea of using the order of the array. Maybe we don't need that option and the "priority mode" is always active?

So if we want to use the method style where possible and the inline style otherwise and the inline function style only where there are multiple event payload parameters: `["method", "inline", "inline-function"]`.

If we want to use the inline function style only: `["inline-function"]`

If we want to use the inline style only (and also report cases where there are multiple event payload parameters): `["inline"]`
Multiple event payload parameters can be merged into one object, so these cases are possible to fix manually.

Specifying only `["method"]` should then be forbidden by the schema because it would otherwise report also cases that are impossible to fix.

Reply:
Thank you for your opinion! I will change this rule without new options.

Reply:
@FloEdelmann I changed the rule. Could you please check again?

Reply:
Sorry, please wait a moment because I thought it should be refactored more.

Reply:
@FloEdelmann I finished refactoring. Please check it again 🙏